Changes in module 'nf-core/pretextmap'
--- modules/nf-core/pretextmap/main.nf
+++ modules/nf-core/pretextmap/main.nf
@@ -3,10 +3,9 @@
     tag "$meta.id"
     label 'process_single'
 
-    conda "bioconda::pretextmap=0.1.9 bioconda::samtools=1.17"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/mulled-v2-f3591ce8609c7b3b33e5715333200aa5c163aa61%3A44321ab4d64f0b6d0c93abbd1406369d1b3da684-0':
-        'biocontainers/mulled-v2-f3591ce8609c7b3b33e5715333200aa5c163aa61:44321ab4d64f0b6d0c93abbd1406369d1b3da684-0' }"
+        'https://depot.galaxyproject.org/singularity/samtools:1.17--h00cdaf9_0' :
+        'biocontainers/samtools:1.17--h00cdaf9_0' }"
 
     input:
     tuple val(meta), path(input)
@@ -20,13 +19,15 @@
     task.ext.when == null || task.ext.when
 
     script:
+    def VERSION = "0.1.9"
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
     def reference = fasta ? "--reference ${fasta}" : ""
+    def pretext_path = "${projectDir}/bin/PretextMap/bin/PretextMap"
 
     """
     if [[ $input == *.pairs.gz ]]; then
-        zcat $input | PretextMap \\
+        zcat $input | ${pretext_path} \\
             $args \\
             -o ${prefix}.pretext
     else
@@ -34,26 +35,28 @@
             view \\
             $reference \\
             -h \\
-            $input | PretextMap \\
+            $input | ${pretext_path} \\
             $args \\
             -o ${prefix}.pretext
     fi
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
-        pretextmap: \$(PretextMap | grep "Version" | sed 's/PretextMap Version //g')
+        pretextmap: $VERSION
         samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' )
     END_VERSIONS
     """
 
     stub:
+    def VERSION = "0.1.9"
     def prefix = task.ext.prefix ?: "${meta.id}"
+    def pretext_path = "${projectDir}/bin/PretextMap/bin/PretextMap"
     """
     touch ${prefix}.pretext
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
-        pretextmap: \$(PretextMap | grep "Version" | sed 's/PretextMap Version //g')
+        pretextmap: $VERSION
         samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//' ))
     END_VERSIONS
     """

************************************************************
