Changes in module 'nf-core/mummer'
--- modules/nf-core/mummer/main.nf
+++ modules/nf-core/mummer/main.nf
@@ -9,7 +9,7 @@
         'biocontainers/mummer:3.23--pl5262h1b792b2_12' }"
 
     input:
-    tuple val(meta), path(ref), path(query)
+    tuple val(meta), path(ref), path('query.fasta') // Staging removes the ability to deal with zipped files here
 
     output:
     tuple val(meta), path("*.coords"), emit: coords
@@ -24,15 +24,16 @@
     def is_compressed_ref = ref.getName().endsWith(".gz") ? true : false
     def fasta_name_ref = ref.getName().replace(".gz", "")
 
-    def is_compressed_query = query.getName().endsWith(".gz") ? true : false
-    def fasta_name_query = query.getName().replace(".gz", "")
+    def query = 'query.fasta'
+    def is_compressed_query = query.endsWith(".gz") ? true : false
+    def fasta_name_query = query.replace(".gz", "")
     def VERSION = '3.23' // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.
     """
     if [ "$is_compressed_ref" == "true" ]; then
         gzip -c -d $ref > $fasta_name_ref
     fi
     if [ "$is_compressed_query" == "true" ]; then
-        gzip -c -d $query > $fasta_name_query
+        gzip -c -d query.fasta > $fasta_name_query
     fi
     mummer \\
         $args \\

************************************************************
