Changes in module 'nf-core/mummer'
--- modules/nf-core/mummer/main.nf
+++ modules/nf-core/mummer/main.nf
@@ -9,7 +9,7 @@
         'biocontainers/mummer:3.23--pl5262h1b792b2_12' }"
 
     input:
-    tuple val(meta), path(ref), path(query)
+    tuple val(meta), path(ref, stageAs: 'ref.fasta'), path(query, stageAs: 'query.fasta') // Staging removes the ability to deal with zipped files here
 
     output:
     tuple val(meta), path("*.coords"), emit: coords

************************************************************
